Modules/_ctypes/cfield.c

```diff
@@ -61,238 +72,146 @@ Py_ssize_t LOW_BIT(Py_ssize_t offset);
-        // allowance for alignment.
-        *pbitofs = - *pfield_size;
+    PyTypeObject *tp = type;
+    ctypes_state *st = get_module_state_by_class(tp);
+    self = (CFieldObject *)tp->tp_alloc(tp, 0);
+    if (!self) {
+        return NULL;
     }
-
-    assert(8 * info->size == *pfield_size);
-
-    self->offset = *poffset - (*pfield_size) / 8;
-    if(is_bitfield) {
-        assert(0 <= (*pfield_size + *pbitofs));
-        assert((*pfield_size + *pbitofs) < info->size * 8);
-        self->size = BUILD_SIZE(bitsize, *pfield_size + *pbitofs);
+    if (PyUnicode_CheckExact(name)) {
+        self->name = Py_NewRef(name);
     } else {
-        self->size = info->size;
+        self->name = PyObject_Str(name);
+        if (!self->name) {
+            goto error;
+        }
     }
-    assert(*pfield_size + *pbitofs <= info->size * 8);
-
-    *pbitofs += bitsize;
-    *psize = *poffset;
-
-    return 0;
-}
 
-PyObject *
-PyCField_FromDesc(ctypes_state *st, PyObject *desc, Py_ssize_t index,
-                Py_ssize_t *pfield_size, Py_ssize_t bitsize,
-                Py_ssize_t *pbitofs, Py_ssize_t *psize, Py_ssize_t *poffset, Py_ssize_t *palign,
-                int pack, int big_endian, LayoutMode layout_mode)
-{
-    PyTypeObject *tp = st->PyCField_Type;
-    CFieldObject* self = (CFieldObject *)tp->tp_alloc(tp, 0);
-    if (self == NULL) {
-        return NULL;
-    }
     StgInfo *info;
-    if (PyStgInfo_FromType(st, desc, &info) < 0) {
-        Py_DECREF(self);
-        return NULL;
+    if (PyStgInfo_FromType(st, proto, &info) < 0) {
+        goto error;
     }
-    if (!info) {
-        PyErr_SetString(PyExc_TypeError,
-                        "has no _stginfo_");
-        Py_DECREF(self);
-        return NULL;
+    if (info == NULL) {
+        PyErr_Format(PyExc_TypeError,
+                     "type of field %R must be a C type", self->name);
+        goto error;
     }
 
-    PyObject* proto = desc;
+    Py_ssize_t bit_size = NUM_BITS(size);
+    if (bit_size) {
+        switch(info->ffi_type_pointer.type) {
+            case FFI_TYPE_UINT8:
+            case FFI_TYPE_UINT16:
+            case FFI_TYPE_UINT32:
+            case FFI_TYPE_SINT64:
+            case FFI_TYPE_UINT64:
+                break;
+
+            case FFI_TYPE_SINT8:
+            case FFI_TYPE_SINT16:
+            case FFI_TYPE_SINT32:
+                if (info->getfunc != _ctypes_get_fielddesc("c")->getfunc
+                    && info->getfunc != _ctypes_get_fielddesc("u")->getfunc)
+                {
+                    break;
+                }
+                _Py_FALLTHROUGH;  /* else fall through */
+            default:
+                PyErr_Format(PyExc_TypeError,
+                             "bit fields not allowed for type %s",
+                             ((PyTypeObject*)proto)->tp_name);
+                goto error;
+            }
+            if (bit_size <= 0 || bit_size > info->size * 8) {
```

Comment:
```suggestion
            if (bit_size < 0 || bit_size > info->size * 8) {
```

You're already in a condition `if (bit_size)`